barcode/src/lib.rs

```diff
@@ -0,0 +1,1231 @@
+];
+
+/// This character is used to separate the various components of a `Barcode`
+pub const SEPARATOR: &str = ":";
+
+/// This type holds information about invalid characters found in our various
+/// identifiers. We provide both the index of the character in the string and
+/// the offending character.
+#[derive(Debug, PartialEq)]
+pub struct InvalidChar {
+    pub index: usize,
+    pub character: char,
+}
+
+impl InvalidChar {
+    pub fn new(index: usize, character: char) -> Self {
+        InvalidChar { index, character }
+    }
+}
+
+impl fmt::Display for InvalidChar {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "Invalid character: {{ index = {}, char = \'{}\' }}",
+            self.index, self.character
+        )
+    }
+}
+
+/// String prefix identifying the first version of the 0xide barcode string
+/// format
+pub const PREFIX_0XV1: &str = "0XV1";
+
+/// String prefix identifying the second version of the 0xide barcode string
+/// format
+pub const PREFIX_0XV2: &str = "0XV2";
+
+/// String prefix identifying the first version of the 0xide platform identity
+/// string format
+pub const PREFIX_PDV1: &str = "PDV1";
+
+/// String prefix identifying the second version of the 0xide platform
+/// identity string format
+pub const PREFIX_PDV2: &str = "PDV2";
+
+/// This type identifies the possible errors encountered while parsing the
+/// prefix extracted from a barcode / identity string
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum PrefixError {
+    #[error("Invalid prefix")]
+    Invalid,
+}
+
+/// A type representing all supported barcode string prefixes
+#[derive(Debug, PartialEq)]
+pub enum Prefix {
+    ZeroXV1,
+    ZeroXV2,
+    PDV1,
+    PDV2,
+}
+
+impl Prefix {
+    /// Get the string representation for the `Prefix` variant
+    pub fn as_str(&self) -> &str {
+        match self {
+            Prefix::ZeroXV1 => PREFIX_0XV1,
+            Prefix::ZeroXV2 => PREFIX_0XV2,
+            Prefix::PDV1 => PREFIX_PDV1,
+            Prefix::PDV2 => PREFIX_PDV2,
+        }
+    }
+}
+
+impl TryFrom<&str> for Prefix {
+    type Error = PrefixError;
+
+    /// Construct an instance of the `Prefix` type for the given string
+    fn try_from(s: &str) -> Result<Prefix, Self::Error> {
+        Ok(match s {
+            PREFIX_0XV1 => Prefix::ZeroXV1,
+            PREFIX_0XV2 => Prefix::ZeroXV2,
+            PREFIX_PDV1 => Prefix::PDV1,
+            PREFIX_PDV2 => Prefix::PDV2,
+            _ => return Err(Self::Error::Invalid),
+        })
+    }
+}
+
+/// This is a utility function to check that each character in the string `s`
+/// is in the dictionary `dict`. If a character is found that is not in `dict`
+/// then we return the offending index & char as `Some(InvalidChar)`.
+/// Otherwise return `None` indicating that all characters in `s` are in `dict`.
+fn is_char_in_dict(s: &str, dict: &[char]) -> Option<InvalidChar> {
```

Comment:
nit: Typically the `is_` is idiomatic for functions that return `bool`. It seems reasonable to also do that when returning an option where `Some` represents `true`. I find it somewhat confusing the way this is named though, because `None` is returned when the char is in the dict. It also is looking at all characters, rather than just one.

I'd suggest changing the name to something like `find_invalid_char_in_dict`.